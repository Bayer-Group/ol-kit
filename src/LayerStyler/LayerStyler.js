import React from 'react'
import PropTypes from 'prop-types'
import StyleManager from 'LayerStyler/StyleManager'
import HeatmapLayer from 'ol/layer/Heatmap'
import olFormatFilterAnd from 'ol/format/filter/And'
import olFormatFilterOr from 'ol/format/filter/Or'
import olFormatFilterEqualTo from 'ol/format/filter/EqualTo'
import olFormatFilterIsLike from 'ol/format/filter/IsLike'
import olFilterFunction from '../classes/FilterFunction'
import ugh from 'ugh'

import escapeRegExp from 'lodash.escaperegexp'

import { addMovementListener, removeMovementListener } from 'Popup'
import { connectToContext } from 'Provider'

/**
 * UI to choose color, stroke, fill etc. for styles and labels on layers
 * @component
 * @category LayerStyler
 */
class LayerStyler extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      attributeValues: [],
      listeners: [],
      whitelistedLayers: props.whitelistedLayers
    }
  }

  componentDidMount () {
    const { map } = this.props
    const layers = map.getLayers()

    // if a layer is added or removed, the list of layers should be updated
    const handleLayersChange = (e) => this.forceUpdate()

    // bind the event listeners
    const onAddKey = layers.on('add', handleLayersChange)
    const onRemoveKey = layers.on('remove', handleLayersChange)

    // make sure the attributes get updated each time the view extent changes
    const listeners = addMovementListener(map, () => this.forceUpdate())

    this.setState({ listeners: [...listeners, onAddKey, onRemoveKey] })
  }

  componentWillUnmount () {
    const { listeners } = this.state

    removeMovementListener(listeners)
  }

  getTitleForLayer = (layer) => {
    return layer?.get?.('title') || layer?.getTypeName?.() || 'untitled layer'
  }

  getAttributesForLayer = layer => {
    if (layer) {
      if (layer.isGeoserverLayer) {
        return layer.getAttributes().sort((a, b) => a.localeCompare(b))
      } else if (layer.isVectorLayer || layer.isVectorTileLayer) {
        return layer.getAttributes().sort((a, b) => a.localeCompare(b))
      }
    }
  }

  getValuesForAttribute = async (layer, attribute) => {
    if (layer) {
      if (layer.isGeoserverLayer) {
        const { typeName } = layer
        const whitelistedLayer = this.state.whitelistedLayers.find(l => l.typename === typeName)
        const opts = whitelistedLayer ? { commaDelimitedAttributes: whitelistedLayer.commaDelimitedAttributes } : {}
        const attributeValues = await layer.fetchValuesForAttribute(this.props.map, attribute, opts)

        this.setState({ attributeValues })
      } else if (layer.isVectorLayer || layer.isVectorTileLayer) {
        const attributeValues = layer.fetchValuesForAttribute(attribute)

        this.setState({ attributeValues })
      }
    }
  }

  getCommaDelimitedAttributesForLayer = (layer) => {
    if (layer && layer.isGeoserverLayer) {
      const { typeName } = layer
      const whitelistedLayer = this.state.whitelistedLayers.find(l => l.typename === typeName)

      return whitelistedLayer ? whitelistedLayer.commaDelimitedAttributes : []
    }
  }

  onFilterChange = (layer, filters) => {
    if (layer && layer.isGeoserverLayer) {
      const { typeName } = layer
      const whitelistedLayer = this.state.whitelistedLayers.find(l => l.typename === typeName)
      const opts = whitelistedLayer ? { commaDelimitedAttributes: whitelistedLayer.commaDelimitedAttributes } : {}

      layer.setWMSFilters(filters, opts)
      layer.setOlFilters(this.setOlFilters(filters, opts))

      // cause a re-render which will re-hydrate the latest styles
      this.forceUpdate()
    }
  }

  setOlFilters = (filters = [], opts) => {
    const olFilters = filters.map(({ attribute, value }) => {
      if (opts.commaDelimitedAttributes && opts.commaDelimitedAttributes.includes(attribute)) {
        // this filter is due to commaDelimitedAttributes, we have to make an OR filter with 3 regex filters
        // the olFilterFunction is our custom filter function due to openlayers not supporting functions as filters
        return new olFormatFilterOr(
          new olFormatFilterIsLike(new olFilterFunction('strMatches', attribute, `^${escapeRegExp(value)}( )??,.*`), true, '*', '.', '!'),
          new olFormatFilterIsLike(new olFilterFunction('strMatches', attribute, `.*,( )??${escapeRegExp(value)}( )??,.*`), true, '*', '.', '!'),
          new olFormatFilterIsLike(new olFilterFunction('strMatches', attribute, `.*,( )??${escapeRegExp(value)}( )??,.*`), true, '*', '.', '!'),
          new olFormatFilterEqualTo(attribute, escapeRegExp(value))
        )
      } else {
        return new olFormatFilterEqualTo(attribute, value)
      }
    })

    if (filters[0]?.logical === 'AND' && filters.length > 1) {
      return new olFormatFilterAnd(...olFilters)
    } else if (filters[0]?.logical === 'OR' && filters.length > 1) {
      return new olFormatFilterOr(...olFilters)
    } else {
      return olFilters[0]
    }
  }

  onDefaultStyleChange = (layer, styles) => {
    if (layer && layer.isGeoserverLayer) {
      layer.setDefaultWMSStyles(styles)
    } else {
      layer.updateDefaultVectorStyles(styles)
    }

    // cause a re-render which will re-hydrate the latest styles
    this.forceUpdate()
  }

  onUserStyleChange = (layer, styles) => {
    if (layer && layer.isGeoserverLayer) {
      layer.setUserWMSStyles(styles)
    } else if (layer && (layer.isVectorLayer || layer.isVectorTileLayer)) {
      layer.setUserVectorStyles(styles)
    }

    // cause a re-render which will re-hydrate the latest styles
    this.forceUpdate()
  }

  onDefaultStyleReset = (layer) => {
    if (layer && layer.isGeoserverLayer) {
      layer.resetDefaultWMSStyles()
    } else if (layer.isVectorLayer || layer.isVectorTileLayer) {
      layer.resetDefaultVectorStyles()
    }

    // cause a re-render which will re-hydrate the latest styles
    this.forceUpdate()
  }

  getValidLayers = () => {
    const { map } = this.props
    const layers = map.getLayers().getArray()
    const validLayers = layers.filter(layer => {
      return !layer.get('_ol_kit_basemap') && (layer.isGeoserverLayer || layer.isVectorLayer || layer.isVectorTileLayer || layer instanceof HeatmapLayer)
    })

    if (layers.length - validLayers.length > 1) {
      ugh.warn('In order to use ManageLayers, the layer must be either an VectorLayer or GeoserverLayer')
    }

    return validLayers
  }

  render () {
    const layers = this.getValidLayers()
    const { attributeValues } = this.state
    const { translations } = this.props

    return (
      <StyleManager
        layers={layers}
        translations={translations}
        filters={layers.map(l => l.isGeoserverLayer && l.getWMSFilters())}
        userStyles={layers.map(l => l.isGeoserverLayer ? l.getUserWMSStyles() : l.getUserVectorStyles?.())}
        defaultStyles={layers.map(l => l.isGeoserverLayer ? l.getDefaultWMSStyles() : l.getDefaultVectorStyles?.())}
        getCommaDelimitedAttributesForLayer={this.getCommaDelimitedAttributesForLayer}
        getTitleForLayer={this.getTitleForLayer}
        getValuesForAttribute={this.getValuesForAttribute}
        getAttributesForLayer={this.getAttributesForLayer}
        attributeValues={attributeValues}
        onFilterChange={this.onFilterChange}
        onUserStyleChange={this.onUserStyleChange}
        onDefaultStyleChange={this.onDefaultStyleChange}
        onDefaultStyleReset={this.onDefaultStyleReset}
        {...this.props} />
    )
  }
}

LayerStyler.defaultProps = {
  whitelistedLayers: []
}

LayerStyler.propTypes = {
  /** Openlayers map object */
  map: PropTypes.object.isRequired,

  /** Object with key/value pairs for translated strings */
  translations: PropTypes.object.isRequired,

  /** An array of layer typenames that will be used for a whitelist */
  whitelistedLayers: PropTypes.array
}

export default connectToContext(LayerStyler)
