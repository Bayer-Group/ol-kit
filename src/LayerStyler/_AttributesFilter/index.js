import React from 'react'
import PropTypes from 'prop-types'
import nanoid from 'nanoid'

import CloseIcon from '@material-ui/icons/Close'
import Select from '@material-ui/core/Select'
import MenuItem from '@material-ui/core/MenuItem'

import { connectToContext } from 'Provider'
import Popover from 'LayerStyler/_Popover'
import Selector from 'LayerStyler/_Selector'
import { ContentContainer, InputContainer, Row, TextContainer, Title } from './styled'

const defaultFilter = {
  logical: 'AND', // determines how to join the filter to other filters
  condition: 'is', // default to 'is' for now
  attribute: '',
  value: ''
}

class AttributesFilter extends React.Component {
  addFilter = () => {
    this.props.onUpdateFilters([...this.props.filters, defaultFilter])
  }

  removeFilter = idx => {
    const newFilters = this.props.filters.filter((f, i) => {
      return i !== idx
    })

    this.props.onUpdateFilters(newFilters)
  }

  // current implementation always updates the logical op for every filter
  updateLogicalOperator (op) {
    const updatedFilters = [...this.props.filters].map(f => ({ ...f, logical: op }))

    this.props.onUpdateFilters(updatedFilters)
  }

  handleSelect = (item, input, idx) => {
    const updateFilter = { ...this.props.filters[idx] }

    if (input === 'attribute') {
      this.props.getValuesForAttribute(item) // parse values based on attribute string -- async
      updateFilter.attribute = item
      updateFilter.value = '' // reset value to empty string since values are calculated per selected attribute
    } else if (input === 'condition') {
      updateFilter.condition = item
    } else if (input === 'value') {
      updateFilter.value = item
    }
    const newFilters = this.props.filters.map((f, i) => {
      return i === idx ? updateFilter : f
    })

    this.props.onUpdateFilters(newFilters)
  }

  render () {
    const { translations, attributes, attributeValues, disabled, filters, filterConditions } = this.props

    const getLogicOperator = (i, filters) => {
      if (i === 0) return translations['_ol_kit.AttributesFilter.where']
      if (i > 1) return filters[0].logical

      return (
        <Select
          value={filters[0].logical}
          onChange={e => this.updateLogicalOperator(e.target.value)}>
          <MenuItem key='AND' value='AND'>{translations['_ol_kit.AttributesFilter.AND']}</MenuItem>
          <MenuItem key='OR'value='OR'>{translations['_ol_kit.AttributesFilter.OR']}</MenuItem>
        </Select>
      )
    }

    return (
      <Popover title={translations['_ol_kit.AttributesFilter.filters'] + (filters.length ? ` (${filters.length})` : '')} disabled={disabled}>
        <ContentContainer>
          <Title>{translations['_ol_kit.AttributesFilter.filters']}</Title>
          {filters.map((filter, i) => {
            return (
              <Row key={nanoid(6)}>
                <CloseIcon onClick={this.removeFilter.bind(this, i)} />
                <TextContainer>
                  {getLogicOperator(i, filters)}
                </TextContainer>
                <InputContainer>
                  <Selector
                    options={attributes}
                    header={`${translations['_ol_kit.AttributesFilter.attribute']} ${i + 1}`}
                    onValueChange={value => this.handleSelect(value, 'attribute', i)}
                    selected={filter.attribute} />
                </InputContainer>
                <InputContainer>
                  <Selector
                    options={filterConditions}
                    header={`${translations['_ol_kit.AttributesFilter.condition']} ${i + 1}`}
                    onValueChange={value => this.handleSelect(value, 'condition', i)}
                    selected={filter.condition} />
                </InputContainer>
                <InputContainer>
                  <Selector
                    options={attributeValues.map(v => `${v}`) /* string interpolation here ensures booleans display properly */}
                    header={`${translations['_ol_kit.AttributesFilter.value']} ${i + 1}`}
                    onValueChange={value => this.handleSelect(value, 'value', i)}
                    selected={filter.value}
                    disabled={!filter.attribute} />
                </InputContainer>
              </Row>
            )
          })}
          <Row><span onClick={this.addFilter}>+ {translations['_ol_kit.AttributesFilter.addFilter']}</span></Row>
        </ContentContainer>
      </Popover>
    )
  }
}

AttributesFilter.propTypes = {
  /** Object with key/value pairs for translated strings */
  translations: PropTypes.object,

  /** Array of attribute strings */
  attributes: PropTypes.array.isRequired,

  /** Disable click on filters button */
  disabled: PropTypes.bool,

  /** Array of filter objects that contain `attribute`: string, `condition`: string, `value`: string, `values`: array */
  filters: PropTypes.array.isRequired,

  /** Optionally override the default filter conditions */
  filterConditions: PropTypes.array,

  /** Callback fired when attribute is selected (can be async) */
  getValuesForAttribute: PropTypes.func.isRequired,

  /** Array of strings representing values for the selected attribute */
  attributeValues: PropTypes.array,

  /** Callback func triggered on filter update with the new filters array */
  onUpdateFilters: PropTypes.func.isRequired
}

AttributesFilter.defaultProps = {
  attributes: [],
  attributeValues: [],
  disabled: false,
  filters: [],
  filterConditions: ['is', 'is not', 'contains'],
  getValuesForAttribute: () => {},
  onUpdateFilters: () => {}
}

export default connectToContext(AttributesFilter)
