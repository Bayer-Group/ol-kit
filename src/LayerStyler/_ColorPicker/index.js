import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { BlockPicker } from 'react-color'
import {
  Container,
  ColorPickerPositioner,
  CurrentColor
} from './styled'

const colors = [
  '#f44336',
  '#9c27b0',
  '#673ab7',
  '#3f51b5',
  '#2196f3',
  '#03a9f4',
  '#00bcd4',
  '#009688',
  '#4caf50',
  '#8bc34a',
  '#cddc39',
  '#ffeb3b',
  '#ffc107',
  '#ff9800',
  '#795548',
  '#607d8b',
  '#000000',
  '#ffffff'
]

class ColorPicker extends Component {
  constructor (props) {
    super(props)

    this.state = {
      colorPickerOpen: false,
      uid: Math.floor((Math.random() * 10000))
    }
  }

  handleColorChange = (opts) => {
    this.props.handleSelect(opts.hex)
  }

  handleOffClick = e => {
    if (this.state.colorPickerOpen) {
      // assume off-click until proven wrong
      let offClick = true

      // loop through elements in clicked path to determine off-click
      e.path.forEach(elem => {
        elem.classList && elem.classList.forEach(className => {
          // uid in className avoids bug where two color pickers can be open at once
          const { uid } = this.state

          // click happened within BlockPicker or CurrentColor so this is not an off-click
          if (className === 'block-picker' || className === `current-color-${uid}`) offClick = false
        })
      })
      // if off-click occured close color picker
      if (offClick) this.setState({ colorPickerOpen: false })
    }
  }

  componentDidMount () {
    document.addEventListener('click', this.handleOffClick)
  }

  componentWillUnmount () {
    document.removeEventListener('click', this.handleOffClick)
  }

  componentDidUpdate (prevProps, prevState) {
    // hacky dom manipulations to force styling on the BlockPicker component
    if (!prevState.colorPickerOpen && this.state.colorPickerOpen) {
      // add border to white swatch
      Array.from(document.querySelectorAll('[title="#ffffff"]')).forEach(elem => {
        elem.style.border = '1px solid #d8d8d8'
      })
      // shrink hex colored header & remove the hex code
      Array.from(document.getElementsByClassName('block-picker')).forEach(elem => {
        const header = elem.children[1]

        header.style.height = '12px'
        header.innerHTML = ''
      })
    }
  }

  render () {
    const { currentColor } = this.props
    const { colorPickerOpen, uid } = this.state

    return (
      <Container>
        <CurrentColor className={`current-color-${uid}`} color={currentColor} onClick={() => this.setState({ colorPickerOpen: !colorPickerOpen })} />
        {colorPickerOpen &&
          <ColorPickerPositioner>
            <BlockPicker color={currentColor} colors={colors} onChangeComplete={this.handleColorChange} />
          </ColorPickerPositioner>
        }
      </Container>
    )
  }
}

ColorPicker.propTypes = {
  /** The currently selected color */
  currentColor: PropTypes.string.isRequired,

  /** Callback function which passes the hex value of the color selected */
  handleSelect: PropTypes.func
}

ColorPicker.defaultProps = {
  currentColor: '#ffffff',
  handleSelect: () => {}
}

export default ColorPicker
