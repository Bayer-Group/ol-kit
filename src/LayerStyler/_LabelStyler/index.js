import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Select from '@material-ui/core/Select'
import MenuItem from '@material-ui/core/MenuItem'

import { connectToContext } from 'Provider'
import ugh from 'ugh'
import ColorPicker from 'LayerStyler/_ColorPicker'

import {
  TopControls,
  AttributeContainer,
  AttributeItem,
  ToggleContainer,
  SymbolizerContainer,
  AttributeHeader,
  Title,
  Button,
  ButtonText,
  Color,
  Outline,
  Size,
  Unit,
  Switch
} from './styled'

const DIVIDER = ':'
const UNITS = {
  ft: ['100ft', '125ft', '150ft', '250ft', '500ft', '750ft', '1000ft', '1250ft', '1500ft', '1750ft', '2000ft'],
  px: ['8px', '10px', '12px', '14px', '16px', '20px', '24px', '30px', '36px', '42px', '48px', '60px', '72px', '96px']
}

class LabelStyler extends Component {
  constructor (props) {
    super(props)

    this.state = {
      uglyChecked: false
    }
  }

  aggregateChanges = (key, value) => {
    const { style, onStylesChange } = this.props

    try {
      const newSymbolizer = {
        ...style.symbolizers[0],
        [key]: value
      }

      style.symbolizers = [newSymbolizer]
    } catch (err) {
      ugh.error(err)

      throw new Error('Failed to aggregate changes to the style')
    }

    onStylesChange([style])
  }

  updateTextSizeUnit = unit => {
    this.aggregateChanges('size', unit === 'ft' ? '250ft' : '16px')
  }

  onAttributeChange = (attr, checked) => {
    const { style, onStylesChange } = this.props
    const newStyle = { ...style }

    try {
      const labelParts = newStyle.symbolizers[0].label.split(DIVIDER)

      let newLabel

      // if the item was previously checked,
      if (checked) {
        newLabel = labelParts.filter(p => p !== `{{${attr}}}`).join(DIVIDER)
      } else {
        // the inner filter here removes empty strings which will add an extra divider
        // which we don't want when the .join() happens
        newLabel = [...labelParts.filter(p => p), `{{${attr}}}`].join(DIVIDER)
      }

      newStyle.symbolizers[0].label = newLabel
    } catch (err) {
      ugh.error(err)

      throw new Error('Failed to modify the symbolizer label')
    }

    onStylesChange([newStyle])
  }

  clearCheckedItems = () => {
    const { style, onStylesChange } = this.props

    try {
      style.symbolizers[0].label = ''
    } catch (err) {
      ugh.error(err)

      throw new Error('There was an issue setting the symbolizer label')
    }

    onStylesChange([style])
  }

  onToggle = () => {
    const { style, onStylesChange } = this.props
    const { uglyChecked } = this.state
    const newStyleSymbolizer = {
      ...style.symbolizers[0],
      spaceAround: uglyChecked ? -5 : 5,
      partials: uglyChecked,
      repeat: uglyChecked ? 1 : 100,
      autoWrap: 100,
      maxDisplacement: 0,
      group: 'no',
      conflictResolution: !uglyChecked,
      goodnessOfFit: 0.1,
      labelAllGroup: false,
      polygonAlign: 'mbr'
    }

    onStylesChange([{ ...style, symbolizers: [newStyleSymbolizer] }])
    this.setState({ uglyChecked: !this.state.uglyChecked })
  }

  render () {
    const { uglyChecked } = this.state
    const { translations, attributes, style } = this.props

    if (!style) return <div>{translations['_ol_kit.LabelStyler.noLabelSupport']}</div>

    const symbolizer = style.symbolizers[0]
    const checkedAttributes = symbolizer.label.split(DIVIDER)
    const sizeUnit = symbolizer.size.slice(-2)

    return (
      <div>
        <TopControls>
          <ToggleContainer>
            <span>{translations['_ol_kit.LabelStyler.enableSmartLabels']}</span>
            <Switch
              checked={uglyChecked}
              onChange={this.onToggle}
              inputProps={{ 'aria-label': 'secondary checkbox' }}
            />
          </ToggleContainer>
          <SymbolizerContainer>
            <Color>
              <Title>{translations['_ol_kit.LabelStyler.color']}</Title>
              <ColorPicker handleSelect={(val) => this.aggregateChanges('color', val)} currentColor={symbolizer.color} />
            </Color>
            <Outline>
              <Title>{translations['_ol_kit.LabelStyler.outline']}</Title>
              <ColorPicker handleSelect={(val) => this.aggregateChanges('haloColor', val)} currentColor={symbolizer.haloColor} />
            </Outline>
            <Size>
              <Title>{translations['_ol_kit.LabelStyler.textHeight']}</Title>
              <Select
                style={{ marginTop: '10px' }}
                value={symbolizer.size}
                onChange={e => this.aggregateChanges('size', e.target.value)}>
                {UNITS[sizeUnit].map(s => {
                  return <MenuItem key={s} value={s}>{s}</MenuItem>
                })}
              </Select>
            </Size>
            <Unit>
              <Title>{translations['_ol_kit.LabelStyler.textUnits']}</Title>
              <Select
                style={{ marginTop: '10px' }}
                value={sizeUnit}
                onChange={e => this.updateTextSizeUnit(e.target.value)}>
                <MenuItem key='px' value={'px'}>{translations['_ol_kit.LabelStyler.pixels']}</MenuItem>
                <MenuItem key='ft' value={'ft'}>{translations['_ol_kit.LabelStyler.feet']}</MenuItem>
              </Select>
            </Unit>
          </SymbolizerContainer>
        </TopControls>
        <AttributeHeader>
          <div>{translations['_ol_kit.LabelStyler.chooseAttrs']}</div>
          {checkedAttributes.length && (
            <Button onClick={this.clearCheckedItems}>
              <ButtonText>{translations['_ol_kit.LabelStyler.clear']}</ButtonText>
            </Button>
          )}
        </AttributeHeader>
        <AttributeContainer data-testid='ManageLayer.attributeContainer'>
          {attributes.map(a => {
            const checked = checkedAttributes.includes(`{{${a}}}`)

            return (
              <AttributeItem key={a} checked={checked} onClick={() => this.onAttributeChange(a, checked)}>
                {a}
              </AttributeItem>
            )
          })}
        </AttributeContainer>
      </div>
    )
  }
}

LabelStyler.propTypes = {
  /** Object with key/value pairs for translated strings */
  translations: PropTypes.object,

  /** array of label styles */
  style: PropTypes.object.isRequired,

  /** array of attributes for the layer */
  attributes: PropTypes.array.isRequired,

  /** callback when label styles change */
  onStylesChange: PropTypes.func.isRequired
}

LabelStyler.defaultProps = {
  style: {},
  attributes: [],
  onStylesChange: () => {}
}

export default connectToContext(LabelStyler)
