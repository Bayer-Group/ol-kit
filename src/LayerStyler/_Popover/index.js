import React from 'react'
import PropTypes from 'prop-types'
import Button from '@material-ui/core/Button'
import MaterialPopover from '@material-ui/core/Popover'
import { withStyles } from '@material-ui/core/styles'

const Popover = withStyles(() => ({
  paper: {
    overflowX: 'visible',
    overflowY: 'visible'
  }
}))(MaterialPopover)

class PopoverBuilder extends React.Component {
  constructor () {
    super()

    this.state = {
      open: false
    }
  }

  handleClickButton = () => {
    this.setState({ open: true })
  }

  handleClose = () => {
    this.setState({ open: false })
  }

  render () {
    const { open } = this.state
    const { children, disabled, title } = this.props

    return (
      <React.Fragment>
        <Button
          disabled={disabled}
          buttonRef={node => {
            this.anchorEl = node
          }}
          variant='contained'
          onClick={this.handleClickButton}>
          {title}
        </Button>
        <Popover
          open={open}
          onClose={this.handleClose}
          anchorEl={this.anchorEl}
          anchorOrigin={{
            vertical: 'top',
            horizontal: 'left'
          }}
          transformOrigin={{
            vertical: 'top',
            horizontal: 'left'
          }}>
          {children}
        </Popover>
      </React.Fragment>
    )
  }
}

PopoverBuilder.propTypes = {
  /** Children to be shown inside the popover when open */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),

  /** If true the popover button will be disabled */
  disabled: PropTypes.bool,

  /** Title to show in the button which triggers the popover */
  title: PropTypes.string
}

export default PopoverBuilder
