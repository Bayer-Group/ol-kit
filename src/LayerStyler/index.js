export { default as LayerStyler } from './LayerStyler'
