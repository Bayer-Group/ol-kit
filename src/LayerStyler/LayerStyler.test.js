import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react'
import translations from 'locales/en'
import olMap from 'ol/Map'
import LayerStyler from './LayerStyler'
import VectorLayer from '../classes/VectorLayer'
import olVectorSource from 'ol/source/Vector'
import olStyle from 'ol/style/Style'
import olStroke from 'ol/style/Stroke'
import olFill from 'ol/style/Fill'
import olCircleStyle from 'ol/style/Circle'
import olFormatGeoJSON from 'ol/format/GeoJSON'

const geojsonObject = {
  type: 'FeatureCollection',
  crs: {
    type: 'name',
    properties: {
      name: 'EPSG:3857'
    }
  },
  features: [{
    type: 'Feature',
    attributes: {
      data: 'testPoint'
    },
    geometry: {
      type: 'Point',
      coordinates: [0, 0]
    }
  }, {
    type: 'Feature',
    attributes: {
      data: 'testLineString'
    },
    geometry: {
      type: 'LineString',
      coordinates: [[4e6, -2e6], [8e6, 2e6]]
    }
  }, {
    type: 'Feature',
    attributes: {
      data: 'testLineString'
    },
    geometry: {
      type: 'LineString',
      coordinates: [[4e6, 2e6], [8e6, -2e6]]
    }
  }, {
    type: 'Feature',
    attributes: {
      data: 'testPolygon'
    },
    geometry: {
      type: 'Polygon',
      coordinates: [[[-5e6, -1e6], [-4e6, 1e6], [-3e6, -1e6]]]
    }
  }, {
    type: 'Feature',
    attributes: {
      data: 'testMultiLineString'
    },
    geometry: {
      type: 'MultiLineString',
      coordinates: [
        [[-1e6, -7.5e5], [-1e6, 7.5e5]],
        [[1e6, -7.5e5], [1e6, 7.5e5]],
        [[-7.5e5, -1e6], [7.5e5, -1e6]],
        [[-7.5e5, 1e6], [7.5e5, 1e6]]
      ]
    }
  }, {
    type: 'Feature',
    geometry: {
      type: 'MultiPolygon',
      coordinates: [
        [[[-5e6, 6e6], [-5e6, 8e6], [-3e6, 8e6], [-3e6, 6e6]]],
        [[[-2e6, 6e6], [-2e6, 8e6], [0, 8e6], [0, 6e6]]],
        [[[1e6, 6e6], [1e6, 8e6], [3e6, 8e6], [3e6, 6e6]]]
      ]
    }
  }, {
    type: 'Feature',
    geometry: {
      type: 'GeometryCollection',
      geometries: [{
        type: 'LineString',
        coordinates: [[-5e6, -5e6], [0, -5e6]]
      }, {
        type: 'Point',
        coordinates: [4e6, -5e6]
      }, {
        type: 'Polygon',
        coordinates: [[[1e6, -6e6], [2e6, -4e6], [3e6, -6e6]]]
      }]
    }
  }]
}

const layerStyle = feature => {
  return new olStyle({
    image: new olCircleStyle({
      radius: 4,
      fill: new olFill({ color: 'rgba(255, 0, 0, 0.4)' }),
      stroke: new olStroke({
        color: 'rgba(255, 0, 0, 0.4)',
        width: 1
      })
    })
  })
}

describe('<LayerStyler />', () => {
  it('should change an ol-kit vector layer\'s style', async done => {
    const map = new olMap()
    const source = new olVectorSource({ features: (new olFormatGeoJSON()).readFeatures(geojsonObject) })
    const title = 'Test Data'
    const vectorLayer = new VectorLayer({ source, title, style: layerStyle })

    source.getFeatures().forEach(f => f.set('testData', `${Math.random() * 100}`))

    map.addLayer(vectorLayer)

    const { getByTestId, getByText } = render(
      <div style={{ height: '100%', width: '500px', backgroundColor: 'white', position: 'absolute' }}>
        {map && <LayerStyler translations={translations} map={map} />}
      </div>, map
    )

    await waitFor(() => expect(map.getLayers().getArray().length).toBe(1))

    await waitFor(() => expect(getByTestId('StyleManager.chooseLayer')))

    fireEvent.change(getByTestId('StyleManager.chooseLayer'), { target: { value: 'Test Data' } })

    fireEvent.click(getByText('Reset Styles'))

    await waitFor(() => expect(getByTestId('StyleManager.chooseLayer').value).toBe('Test Data'))

    await waitFor(() => expect(getByText('Default Styles (1)')).toBeTruthy())

    await waitFor(() => expect(getByText('Custom Styles (0)')).toBeTruthy())

    expect(vectorLayer.getUserVectorStyles().length).toBe(0)

    const addStyle = getByTestId('LayerStyler.addStyle')

    fireEvent.click(addStyle)

    const attributeSelector = getByTestId('StyleGroup.attributeSelector')

    fireEvent.click(attributeSelector)

    fireEvent.change(attributeSelector, { target: { value: 'testData' } })

    await waitFor(() => expect(getByText('Custom Styles (1)')).toBeTruthy())

    await waitFor(() => expect(attributeSelector.value).toBe('testData'))

    await waitFor(() => expect(vectorLayer.getUserVectorStyles()).toMatchSnapshot())

    done()
  })
  it('should add a label to a layer\'s style', async done => {
    const map = new olMap()
    const source = new olVectorSource({ features: (new olFormatGeoJSON()).readFeatures(geojsonObject) })
    const title = 'Test Data'
    const vectorLayer = new VectorLayer({ source, title, style: layerStyle })

    source.getFeatures().forEach(f => f.set('testData', `${Math.random() * 100}`))

    map.addLayer(vectorLayer)

    const { getByTestId, getByText } = render(
      <div style={{ height: '100%', width: '500px', backgroundColor: 'white', position: 'absolute' }}>
        {map && <LayerStyler translations={translations} map={map} />}
      </div>, map
    )

    await waitFor(() => expect(map.getLayers().getArray().length).toBe(1))

    fireEvent.click(getByTestId('StyleManager.labelTab'))

    await waitFor(() => expect(getByText('testData')).toBeTruthy())

    fireEvent.click(getByText('testData'))

    await waitFor(() => expect(vectorLayer.getUserVectorStyles()[0].symbolizers[0].label).toBe('{{testData}}'))
    await waitFor(() => expect(vectorLayer.getUserVectorStyles()).toMatchSnapshot())

    done()
  })
})
