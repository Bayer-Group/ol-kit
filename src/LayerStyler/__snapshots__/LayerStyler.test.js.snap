// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LayerStyler /> should add a label to a layer's style 1`] = `
Array [
  Object {
    "name": "New Label Style",
    "symbolizers": Array [
      Object {
        "LabelPlacement": Array [
          Object {
            "PointPlacement": Object {
              "AnchorPoint": Object {
                "AnchorPointX": 0.5,
                "AnchorPointY": 0.5,
              },
              "Displacement": Object {
                "DisplacementX": 0,
                "DisplacementY": 0,
              },
              "Rotation": 0,
            },
          },
        ],
        "color": "#000",
        "font": Array [
          "Arial",
        ],
        "fontStyle": "normal",
        "fontWeight": "bold",
        "haloColor": "#fff",
        "haloWidth": 1,
        "kind": "Text",
        "label": "{{testData}}",
        "size": "16px",
      },
    ],
  },
]
`;

exports[`<LayerStyler /> should change an ol-kit vector layer's style 1`] = `
Array [
  Object {
    "filter": Array [
      "&&",
      Array [
        "==",
        "",
        "",
      ],
    ],
    "name": "New Auto Style",
    "symbolizers": Array [
      Object {
        "color": "#ffffff",
        "kind": "Mark",
        "radius": 4,
        "strokeColor": "#000000",
        "strokeWidth": 2,
        "wellKnownName": "Circle",
      },
      Object {
        "color": "#ffffff",
        "kind": "Line",
        "width": 7,
      },
      Object {
        "color": "#ffffff",
        "kind": "Fill",
        "opacity": 1,
        "outlineColor": "#000000",
        "outlineWidth": 4,
      },
    ],
  },
]
`;
